File: hahlw/array/39.组合总和.cpp
/*
 * @lc app=leetcode.cn id=39 lang=cpp
 *
 * [39] 组合总和
 */
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>>res;
        vector<int>cur;
        helper(candidates,res,cur,0,target,0);
        return res;
    }
    void helper(vector<int>&candidates,vector<vector<int>>&res,vector<int>&cur,int ct,int target,int start){
        if(ct==target){
            res.push_back(cur);
            return ;
        }
        if(ct>target)return ;
        //允许重复。
        for(int i=start;i<candidates.size();i++){
            cur.push_back(candidates[i]);
            helper(candidates,res,cur,ct+candidates[i],target,i);
            cur.pop_back();
        }
        //不允许重复。
        /*
        for(int i=start;i<candidates.size();i++){
            cur.push_back(candidates[i]);
            helper(candidates,res,cur,ct+candidates[i],target,i+1);
            cur.pop_back();
        }
         */
        
    }
};

File: Juniper/39.组合总和.py
#
# @lc app=leetcode.cn id=39 lang=python3
#
# [39] 组合总和
#
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        lens = len(candidates)
        temp = []
        for i in range(lens):
            if self.check(candidates[i], target):
                if candidates[i] == target:
                    temp.append([candidates[i]])
                else:
                    t = self.match(candidates[i:], target - candidates[i])
                    if t:
                        for p in t:
                            temp.append([candidates[i]]+p)

                    else:
                        continue
        return temp
                
    def match(self,candidates,target):
        lens =  len(candidates)
        temp = []
        for i in range(lens):
            if self.check(candidates[i], target):
                if not target == candidates[i]:
                    t = self.match(candidates[i:], target - candidates[i])
                    if t:
                        for p in t :
                            temp.append([candidates[i]]+p)
                    else:
                        continue
                    #     return []
                else:
                    temp.append([candidates[i]])
        return temp



    

    def check(self, num, target):
        if num > target:
            return False
        else :
            return True

